File: src/cpp/publisher.cpp
#include "publisher.h"

#include <thread>
#include <chrono>

#include <zmq.hpp>
#include <boost/algorithm/string.hpp>

#include "error.h"

using namespace bjcomm;
using namespace boost::algorithm;

Publisher::Publisher(std::string address): 
    _address(address) {}
    
bool Publisher::start(){
    //if already started return
    if(_running) return false;
    
    _socket = new zmq::socket_t (_context, ZMQ_PUB);
    
    //try to connect to socket
    try{
        _socket->connect(("ipc://"+BJCOMM_COMMON_PATH+_address).c_str());
    }catch(zmq::error_t &e){
        delete _socket;
        _socket = 0;
        
        return false;
    }
    
    _running = true;
    return true;
}

void Publisher::send(const Message &msg){
    //throw exception if trying to publish on closed channel
    if(!_running) throw CommunicationError("Trying to send message on channel that is not running!");
        
    //get message
    std::string str = msg.getType()+" "+msg.getData();
    trim(str);
    
    //send message
    zmq::message_t zmq_msg(str.size());
    memcpy (zmq_msg.data(), str.data(), str.size());
    _socket->send(zmq_msg);
}

File: src/java/Publisher.java
/*
 * Copyright (c) 2015 Blue Jay Eindhoven
 */
package nl.bluejayeindhoven.bjcomm;

import java.io.File;

import org.zeromq.ZMQ;
import org.zeromq.ZMQException;

import nl.bluejayeindhoven.bjcomm.Message;
import nl.bluejayeindhoven.bjcomm.CommunicationError;

/**
 * Interface that send a message asynchronously on a socket it connects
 *
 * @see Subscriber
 * @author Koen Wolters
 */
public class Publisher extends CommunicationInterface{
    /**
     * Address of the socket
     */
    private String address;
    
    /**
     * Constructs a publisher that shall communicate over the supplied address
     *
     * @param address address of the socket
     */
    public Publisher(String address){
        this.address = address;
    }
    
    /**
     * {@inheritDoc}
     *
     * Creates a socket and connects to it (not blocking if not binded yet)
     */
    public boolean start(){
        //if already started return false
        if(running) return false;
        
        socket = context.socket(ZMQ.PUB);
        
        try{
			new File(COMMON_PATH+address).mkdirs();
			
            socket.connect("ipc://"+COMMON_PATH+address);
        }catch(ZMQException e){
            socket.close();
            socket = null;
            return false;
        }
        
        running = true;
        return true;
    }
    
    /**
     * Send a message over the channel
     *
     * @param msg message to send over the channel
     *
     * @throws CommunicationError if interface is not started
     */
    public void send(Message msg){
        //throw exception if trying to send message on closed channel
        if(!running) throw new CommunicationError("Trying to send message on channel that is not running!");
        
        //send message
        String str = msg.getType()+" "+msg.getData();
        socket.send(str);
    }
}